p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,314 @@
+)
+
+// TODO Should we have options for these ?
+const (
+	protocol         = "/libp2p/holepunch/1.0.0"
+	maxMsgSize       = 4 * 1024 // 4K
+	holePunchTimeout = 1 * time.Minute
+	dialTimeout      = 5 * time.Second
+	maxRetries       = 5
+	retryWait        = 2 * time.Second
+)
+
+var (
+	log = logging.Logger("p2p-holepunch")
+)
+
+// TODO Find a better name for this protocol.
+// HolePunchService is used to make direct connections with a peer via hole-punching.
+type HolePunchService struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	ids  *identify.IDService
+	host host.Host
+
+	// ensure we shutdown ONLY once
+	closeSync sync.Once
+	refCount  sync.WaitGroup
+}
+
+// NewHolePunchService creates a new service that can be used for hole punching
+func NewHolePunchService(h host.Host, ids *identify.IDService) (*HolePunchService, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	hs := &HolePunchService{ctx: ctx, ctxCancel: cancel, host: h, ids: ids}
+
+	h.SetStreamHandler(protocol, hs.handleNewStream)
+	h.Network().Notify((*netNotifiee)(hs))
+	return hs, nil
+}
+
+// Close closes the Hole Punch Service.
+func (hs *HolePunchService) Close() error {
+	hs.closeSync.Do(func() {
+		hs.ctxCancel()
+		hs.refCount.Wait()
+	})
+
+	return nil
+}
+
+// attempts to make a direct connection with the remote peer of `relayConn` by co-ordinating a hole punch over
+// the given relay connection `relayConn`.
+func (hs *HolePunchService) holePunch(relayConn network.Conn) {
+	rp := relayConn.RemotePeer()
+
+	// short-circuit hole punching if a direct dial works.
+	// attempt a direct connection ONLY if we have a public address for the remote peer
+	for _, a := range hs.host.Peerstore().Addrs(rp) {
+		if manet.IsPublicAddr(a) && !isRelayAddress(a) {
+			forceDirectConnCtx := network.WithForceDirectDial(hs.ctx, "hole-punching")
+			dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+			defer cancel()
+			if err := hs.host.Connect(dialCtx, peer.AddrInfo{ID: rp}); err == nil {
+				log.Debugf("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+				return
+			}
+			break
+		}
+	}
+
+	// hole punch
+	hpCtx := network.WithUseTransient(hs.ctx, "hole-punch")
+	sCtx := network.WithNoDial(hpCtx, "hole-punch")
+	s, err := hs.host.NewStream(sCtx, rp, protocol)
+	if err != nil {
+		log.Errorf("failed to open hole-punching stream with peer %s, err: %s", rp, err)
+		return
+	}
+	log.Infof("will attempt hole punch with peer %s", rp.Pretty())
+	_ = s.SetDeadline(time.Now().Add(holePunchTimeout))
+	w := protoio.NewDelimitedWriter(s)
+
+	// send a CONNECT and start RTT measurement
+	msg := new(pb.HolePunch)
+	msg.Type = pb.HolePunch_CONNECT.Enum()
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	if err := w.WriteMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to send hole punch CONNECT, err: %s", err)
+		return
+	}
+	tstart := time.Now()
```

Comment:
the start of the RTT measurement has to move before writing the message I think.